app/src/main/java/com/esri/arcgisruntime/opensourceapps/datacollection/views/popup/PopupView.kt

```diff
@@ -71,28 +123,178 @@ class PopupView : FrameLayout {
+    }
+
+    /**
+     * Saves the popup by applying changes to the feature service.
+     */
+    private fun savePopup() {
+        val editingErrorFuture: ListenableFuture<ArcGISRuntimeException> = popupManager.finishEditingAsync()
+        editingErrorFuture.addDoneListener {
+            try {
+                editingErrorFuture.get()
+
+                val feature: Feature = popupManager.popup.geoElement as Feature
+                val featureTable: FeatureTable = feature.featureTable
+                if (featureTable is ServiceFeatureTable) {
+                    val applyEditsFuture: ListenableFuture<List<FeatureEditResult>> =
+                        featureTable.applyEditsAsync()
+                    applyEditsFuture.addDoneListener {
+                        try {
+                            val featureEditResults: List<FeatureEditResult> = applyEditsFuture.get()
+                            if (checkEditResultForErrors(featureEditResults)) {
+                                showSnackbarMessage("There was a problem saving the feature")
+                            } else {
+                                showSnackbarMessage("Feature saved successfully")
+                            }
+
+                        } catch (ie: InterruptedException) {
+                            Logger.e(
+                                "Exception encountered when saving popup, " + Util.getExceptionMessageAndCause(
+                                    ie
+                                )
+                            )
+                        } catch (ee: ExecutionException) {
+                            Logger.e(
+                                "Exception encountered when saving popup, " + Util.getExceptionMessageAndCause(
+                                    ee
+                                )
+                            )
+                        }
+                    }
+                }
+            } catch (ie: InterruptedException) {
+                Logger.e(
+                    "Exception encountered when saving popup, " + Util.getExceptionMessageAndCause(
+                        ie
+                    )
                 )
-            )
+            } catch (ee: ExecutionException) {
+                Logger.e(
+                    "Exception encountered when saving popup, " + Util.getExceptionMessageAndCause(
+                        ee
+                    )
+                )
+            }
         }
-        return popupDisplayFieldsWithValues
+    }
+
+    /**
+     * Checks for errors in FeatureEditResults obtained from applying edits to service feature table.
+     */
+    private fun checkEditResultForErrors(featureEditResults: List<FeatureEditResult>): Boolean {
+        var errorEncountered = false
+        for (result in featureEditResults) {
+            if (result.hasCompletedWithErrors()) {
+                errorEncountered = true
+                break
+            }
+        }
+        return errorEncountered
```

Comment:
this doesn't change anything, but seems like a good spot for some functional programming

```suggestion
       featureEditResults
            .filter { result ->
                result.hasCompletedWithErrors()
            }
            .first { return true }

        return false
```

Reply:
updated to use any()